File: backend/email/email_service.py
"""
Email service with AWS SES and SendGrid support.

Supports two email providers:
1. AWS SES (default, recommended for cost)
2. SendGrid (alternative, easier setup)

Configuration via environment variables:
- EMAIL_PROVIDER: "ses" or "sendgrid" (default: "ses")
- EMAIL_FROM: Sender email address (e.g., "noreply@lightengine.io")
- EMAIL_FROM_NAME: Sender name (e.g., "Light Engine")

For AWS SES:
- AWS_REGION: AWS region (e.g., "us-east-1")
- AWS credentials configured via ~/.aws/credentials or environment variables

For SendGrid:
- SENDGRID_API_KEY: SendGrid API key

Usage:
    from backend.email import EmailService, get_password_reset_template
    
    email_service = EmailService()
    template = get_password_reset_template(
        user_name="John Doe",
        reset_url="https://app.lightengine.io/reset-password?token=abc123"
    )
    
    email_service.send_email(
        to_email="user@example.com",
        subject=template.subject,
        html_body=template.html_body,
        text_body=template.text_body
    )
"""

import os
import logging
from typing import Optional, Dict, Any
from dataclasses import dataclass

logger = logging.getLogger(__name__)


@dataclass
class EmailTemplate:
    """Email template with subject and body."""
    subject: str
    html_body: str
    text_body: str


class EmailService:
    """
    Email service with multiple provider support.
    
    Automatically selects provider based on EMAIL_PROVIDER environment variable.
    Falls back to console logging if no provider is configured.
    """
    
    def __init__(self):
        """Initialize email service with configured provider."""
        self.provider = os.getenv("EMAIL_PROVIDER", "ses").lower()
        self.from_email = os.getenv("EMAIL_FROM", "noreply@lightengine.io")
        self.from_name = os.getenv("EMAIL_FROM_NAME", "Light Engine")
        self.enabled = os.getenv("EMAIL_ENABLED", "false").lower() == "true"
        
        if not self.enabled:
            logger.info("📧 Email service DISABLED - emails will be logged to console")
            return
        
        if self.provider == "ses":
            self._init_ses()
        elif self.provider == "sendgrid":
            self._init_sendgrid()
        else:
            logger.warning(f"Unknown email provider: {self.provider}. Emails will be logged to console.")
            self.enabled = False
    
    def _init_ses(self):
        """Initialize AWS SES client."""
        try:
            import boto3
            from botocore.exceptions import ClientError
            
            region = os.getenv("AWS_REGION", "us-east-1")
            self.ses_client = boto3.client("ses", region_name=region)
            
            # Verify SES is accessible
            self.ses_client.get_send_quota()
            logger.info(f"✅ AWS SES email service initialized (region: {region})")
            
        except ImportError:
            logger.error("❌ boto3 not installed. Install with: pip install boto3")
            self.enabled = False
        except Exception as e:
            logger.error(f"❌ Failed to initialize AWS SES: {e}")
            self.enabled = False
    
    def _init_sendgrid(self):
        """Initialize SendGrid client."""
        try:
            from sendgrid import SendGridAPIClient
            
            api_key = os.getenv("SENDGRID_API_KEY")
            if not api_key:
                raise ValueError("SENDGRID_API_KEY environment variable not set")
            
            self.sendgrid_client = SendGridAPIClient(api_key)
            logger.info("✅ SendGrid email service initialized")
            
        except ImportError:
            logger.error("❌ sendgrid not installed. Install with: pip install sendgrid")
            self.enabled = False
        except Exception as e:
            logger.error(f"❌ Failed to initialize SendGrid: {e}")
            self.enabled = False
    
    def send_email(
        self,
        to_email: str,
        subject: str,
        html_body: str,
        text_body: Optional[str] = None,
        reply_to: Optional[str] = None,
        **kwargs
    ) -> bool:
        """
        Send email via configured provider.
        
        Args:
            to_email: Recipient email address
            subject: Email subject line
            html_body: HTML email body
            text_body: Plain text email body (fallback)
            reply_to: Reply-to email address
            **kwargs: Additional provider-specific parameters
        
        Returns:
            True if email sent successfully, False otherwise
        """
        if not self.enabled:
            logger.info(f"""
📧 EMAIL (Console Mode - EMAIL_ENABLED=false)
━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
To: {to_email}
From: {self.from_name} <{self.from_email}>
Subject: {subject}
━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━

{text_body or html_body}
━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
""")
            return True
        
        try:
            if self.provider == "ses":
                return self._send_via_ses(to_email, subject, html_body, text_body, reply_to)
            elif self.provider == "sendgrid":
                return self._send_via_sendgrid(to_email, subject, html_body, text_body, reply_to)
            else:
                logger.error(f"No email provider configured")
                return False
                
        except Exception as e:
            logger.error(f"❌ Failed to send email to {to_email}: {e}")
            return False
    
    def _send_via_ses(
        self,
        to_email: str,
        subject: str,
        html_body: str,
        text_body: Optional[str],
        reply_to: Optional[str]
    ) -> bool:
        """Send email via AWS SES."""
        try:
            from botocore.exceptions import ClientError
            
            message = {
                "Subject": {"Data": subject, "Charset": "UTF-8"},
                "Body": {}
            }
            
            if html_body:
                message["Body"]["Html"] = {"Data": html_body, "Charset": "UTF-8"}
            
            if text_body:
                message["Body"]["Text"] = {"Data": text_body, "Charset": "UTF-8"}
            
            send_params = {
                "Source": f"{self.from_name} <{self.from_email}>",
                "Destination": {"ToAddresses": [to_email]},
                "Message": message,
            }
            
            if reply_to:
                send_params["ReplyToAddresses"] = [reply_to]
            
            response = self.ses_client.send_email(**send_params)
            logger.info(f"✅ Email sent via SES to {to_email} (MessageId: {response['MessageId']})")
            return True
            
        except ClientError as e:
            error_code = e.response['Error']['Code']
            if error_code == 'MessageRejected':
                logger.error(f"❌ SES rejected email to {to_email}: {e.response['Error']['Message']}")
            elif error_code == 'MailFromDomainNotVerifiedException':
                logger.error(f"❌ SES sender domain not verified: {self.from_email}")
            else:
                logger.error(f"❌ SES error sending to {to_email}: {e}")
            return False
    
    def _send_via_sendgrid(
        self,
        to_email: str,
        subject: str,
        html_body: str,
        text_body: Optional[str],
        reply_to: Optional[str]
    ) -> bool:
        """Send email via SendGrid."""
        try:
            from sendgrid.helpers.mail import Mail, Email, To, Content
            
            from_email = Email(self.from_email, self.from_name)
            to_email_obj = To(to_email)
            
            mail = Mail(
                from_email=from_email,
                to_emails=to_email_obj,
                subject=subject,
            )
            
            if text_body:
                mail.add_content(Content("text/plain", text_body))
            
            if html_body:
                mail.add_content(Content("text/html", html_body))
            
            if reply_to:
                mail.reply_to = Email(reply_to)
            
            response = self.sendgrid_client.send(mail)
            logger.info(f"✅ Email sent via SendGrid to {to_email} (status: {response.status_code})")
            return True
            
        except Exception as e:
            logger.error(f"❌ SendGrid error sending to {to_email}: {e}")
            return False
    
    def verify_sender_email(self, email: str) -> bool:
        """
        Verify sender email address with provider.
        
        For AWS SES: Initiates verification process
        For SendGrid: Verifies sender identity
        
        Args:
            email: Email address to verify
        
        Returns:
            True if verification initiated/successful
        """
        if not self.enabled:
            logger.info(f"📧 Email disabled - skipping verification for {email}")
            return False
        
        try:
            if self.provider == "ses":
                self.ses_client.verify_email_identity(EmailAddress=email)
                logger.info(f"✅ SES verification email sent to {email}")
                return True
            elif self.provider == "sendgrid":
                logger.info("⚠️  SendGrid verification must be done via dashboard")
                logger.info("   Visit: https://app.sendgrid.com/settings/sender_auth")
                return False
            else:
                return False
                
        except Exception as e:
            logger.error(f"❌ Failed to verify sender email {email}: {e}")
            return False

File: server/services/email-service.js
/**
 * Unified Email Service
 * Supports AWS SES (preferred), SendGrid, and SMTP fallback
 * 
 * Environment Variables:
 * - EMAIL_PROVIDER: 'ses' | 'sendgrid' | 'smtp' (default: 'ses')
 * - AWS_REGION: AWS region for SES (default: 'us-east-1')
 * - AWS_ACCESS_KEY_ID: AWS credentials (optional if using IAM role)
 * - AWS_SECRET_ACCESS_KEY: AWS credentials (optional if using IAM role)
 * - SENDGRID_API_KEY: SendGrid API key
 * - SMTP_HOST: SMTP server hostname
 * - SMTP_PORT: SMTP server port (default: 587)
 * - SMTP_USER: SMTP username
 * - SMTP_PASS: SMTP password
 * - EMAIL_FROM: Default sender email address
 * - EMAIL_FROM_NAME: Default sender name
 */

import { SESClient, SendEmailCommand } from '@aws-sdk/client-ses';
import nodemailer from 'nodemailer';

class EmailService {
  constructor() {
    this.provider = process.env.EMAIL_PROVIDER || 'ses';
    this.fromEmail = process.env.EMAIL_FROM || 'noreply@farm.local';
    this.fromName = process.env.EMAIL_FROM_NAME || 'Light Engine Foxtrot';
    this.initialized = false;
    this.transporter = null;
    this.sesClient = null;
  }

  /**
   * Initialize email service based on configured provider
   */
  async initialize() {
    if (this.initialized) return;

    try {
      switch (this.provider) {
        case 'ses':
          await this.initializeSES();
          break;
        case 'sendgrid':
          await this.initializeSendGrid();
          break;
        case 'smtp':
          await this.initializeSMTP();
          break;
        default:
          throw new Error(`Unknown email provider: ${this.provider}`);
      }
      
      this.initialized = true;
      console.log(`[email-service] Initialized with provider: ${this.provider}`);
    } catch (error) {
      console.error('[email-service] Initialization failed:', error.message);
      throw error;
    }
  }

  /**
   * Initialize AWS SES client
   */
  async initializeSES() {
    const region = process.env.AWS_REGION || 'us-east-1';
    
    const config = { region };
    
    // Use explicit credentials if provided, otherwise rely on IAM role
    if (process.env.AWS_ACCESS_KEY_ID && process.env.AWS_SECRET_ACCESS_KEY) {
      config.credentials = {
        accessKeyId: process.env.AWS_ACCESS_KEY_ID,
        secretAccessKey: process.env.AWS_SECRET_ACCESS_KEY
      };
    }
    
    this.sesClient = new SESClient(config);
    console.log(`[email-service] AWS SES configured for region: ${region}`);
  }

  /**
   * Initialize SendGrid via nodemailer
   */
  async initializeSendGrid() {
    const apiKey = process.env.SENDGRID_API_KEY;
    
    if (!apiKey) {
      throw new Error('SENDGRID_API_KEY is required for SendGrid provider');
    }
    
    this.transporter = nodemailer.createTransport({
      host: 'smtp.sendgrid.net',
      port: 587,
      secure: false,
      auth: {
        user: 'apikey',
        pass: apiKey
      }
    });
    
    console.log('[email-service] SendGrid SMTP configured');
  }

  /**
   * Initialize generic SMTP
   */
  async initializeSMTP() {
    const host = process.env.SMTP_HOST;
    const port = parseInt(process.env.SMTP_PORT || '587');
    const user = process.env.SMTP_USER;
    const pass = process.env.SMTP_PASS;
    
    if (!host || !user || !pass) {
      throw new Error('SMTP_HOST, SMTP_USER, and SMTP_PASS are required for SMTP provider');
    }
    
    this.transporter = nodemailer.createTransport({
      host,
      port,
      secure: port === 465,
      auth: { user, pass }
    });
    
    console.log(`[email-service] SMTP configured: ${host}:${port}`);
  }

  /**
   * Send email via AWS SES
   */
  async sendViaSES({ to, subject, html, text }) {
    const command = new SendEmailCommand({
      Source: `${this.fromName} <${this.fromEmail}>`,
      Destination: {
        ToAddresses: Array.isArray(to) ? to : [to]
      },
      Message: {
        Subject: { Data: subject },
        Body: {
          Html: { Data: html },
          Text: { Data: text || this.stripHtml(html) }
        }
      }
    });
    
    const response = await this.sesClient.send(command);
    return {
      success: true,
      messageId: response.MessageId,
      provider: 'ses'
    };
  }

  /**
   * Send email via nodemailer (SendGrid or SMTP)
   */
  async sendViaNodemailer({ to, subject, html, text, cc, bcc, attachments }) {
    const info = await this.transporter.sendMail({
      from: `${this.fromName} <${this.fromEmail}>`,
      to: Array.isArray(to) ? to.join(', ') : to,
      cc: cc ? (Array.isArray(cc) ? cc.join(', ') : cc) : undefined,
      bcc: bcc ? (Array.isArray(bcc) ? bcc.join(', ') : bcc) : undefined,
      subject,
      text: text || this.stripHtml(html),
      html,
      attachments
    });
    
    return {
      success: true,
      messageId: info.messageId,
      provider: this.provider
    };
  }

  /**
   * Send email with automatic provider selection
   * 
   * @param {Object} options - Email options
   * @param {string|string[]} options.to - Recipient email(s)
   * @param {string} options.subject - Email subject
   * @param {string} options.html - HTML content
   * @param {string} [options.text] - Plain text content (auto-generated if not provided)
   * @param {string|string[]} [options.cc] - CC recipients
   * @param {string|string[]} [options.bcc] - BCC recipients
   * @param {Array} [options.attachments] - Nodemailer attachments array
   * @returns {Promise<Object>} Result with success, messageId, and provider
   */
  async sendEmail(options) {
    if (!this.initialized) {
      await this.initialize();
    }
    
    const { to, subject, html, text, cc, bcc, attachments } = options;
    
    // Validate required fields
    if (!to || !subject || !html) {
      throw new Error('to, subject, and html are required');
    }
    
    try {
      let result;
      
      if (this.provider === 'ses') {
        result = await this.sendViaSES({ to, subject, html, text });
      } else {
        result = await this.sendViaNodemailer({ to, subject, html, text, cc, bcc, attachments });
      }
      
      console.log(`[email-service] Email sent via ${result.provider} to ${to}`, {
        messageId: result.messageId,
        subject
      });
      
      return result;
    } catch (error) {
      console.error('[email-service] Failed to send email:', error.message);
      throw error;
    }
  }

  /**
   * Send order confirmation email
   */
  async sendOrderConfirmation({ to, orderId, customerName, items, total, farmName }) {
    const html = `
      <!DOCTYPE html>
      <html>
      <head>
        <style>
          body { font-family: Arial, sans-serif; line-height: 1.6; color: #333; }
          .container { max-width: 600px; margin: 0 auto; padding: 20px; }
          .header { background: #2e7d32; color: white; padding: 20px; text-align: center; }
          .content { background: #f9f9f9; padding: 20px; }
          .order-details { background: white; padding: 15px; margin: 15px 0; border-radius: 5px; }
          .item { display: flex; justify-content: space-between; padding: 10px 0; border-bottom: 1px solid #eee; }
          .total { font-size: 1.2em; font-weight: bold; color: #2e7d32; text-align: right; margin-top: 15px; }
          .footer { text-align: center; color: #666; font-size: 0.9em; margin-top: 20px; }
        </style>
      </head>
      <body>
        <div class="container">
          <div class="header">
            <h1>Order Confirmation</h1>
          </div>
          <div class="content">
            <p>Hi ${customerName},</p>
            <p>Thank you for your order from ${farmName}!</p>
            
            <div class="order-details">
              <h2>Order #${orderId}</h2>
              ${items.map(item => `
                <div class="item">
                  <span>${item.name} x ${item.quantity}</span>
                  <span>$${item.price.toFixed(2)}</span>
                </div>
              `).join('')}
              <div class="total">Total: $${total.toFixed(2)}</div>
            </div>
            
            <p>You will receive a notification when your order is ready for pickup.</p>
          </div>
          <div class="footer">
            <p>${farmName} - Light Engine Foxtrot</p>
          </div>
        </div>
      </body>
      </html>
    `;
    
    return this.sendEmail({
      to,
      subject: `Order Confirmation #${orderId} - ${farmName}`,
      html
    });
  }

  /**
   * Send recall notification email
   */
  async sendRecallNotification({ to, lotCode, productName, reason, customerName }) {
    const html = `
      <!DOCTYPE html>
      <html>
      <head>
        <style>
          body { font-family: Arial, sans-serif; line-height: 1.6; color: #333; }
          .container { max-width: 600px; margin: 0 auto; padding: 20px; }
          .header { background: #d32f2f; color: white; padding: 20px; text-align: center; }
          .alert { background: #fff3cd; border-left: 4px solid #ffc107; padding: 15px; margin: 15px 0; }
          .content { background: #f9f9f9; padding: 20px; }
          .footer { text-align: center; color: #666; font-size: 0.9em; margin-top: 20px; }
        </style>
      </head>
      <body>
        <div class="container">
          <div class="header">
            <h1>Product Recall Notice</h1>
          </div>
          <div class="content">
            <p>Dear ${customerName},</p>
            
            <div class="alert">
              <strong>IMPORTANT:</strong> This is a mandatory recall notification.
            </div>
            
            <p>We are issuing a voluntary recall for the following product you purchased:</p>
            
            <ul>
              <li><strong>Product:</strong> ${productName}</li>
              <li><strong>Lot Code:</strong> ${lotCode}</li>
              <li><strong>Reason:</strong> ${reason}</li>
            </ul>
            
            <p><strong>What you should do:</strong></p>
            <ol>
              <li>Do not consume this product</li>
              <li>Return the product to our farm for a full refund</li>
              <li>Contact us if you have any health concerns</li>
            </ol>
            
            <p>We apologize for any inconvenience and take food safety very seriously.</p>
          </div>
          <div class="footer">
            <p>Light Engine Foxtrot - Farm Management System</p>
          </div>
        </div>
      </body>
      </html>
    `;
    
    return this.sendEmail({
      to,
      subject: `URGENT: Product Recall Notice - ${productName}`,
      html
    });
  }

  /**
   * Strip HTML tags for plain text fallback
   */
  stripHtml(html) {
    return html
      .replace(/<style[^>]*>.*?<\/style>/gi, '')
      .replace(/<[^>]+>/g, '')
      .replace(/\s+/g, ' ')
      .trim();
  }

  /**
   * Test email configuration
   */
  async testConnection() {
    if (!this.initialized) {
      await this.initialize();
    }
    
    if (this.transporter) {
      return await this.transporter.verify();
    }
    
    // For SES, just check if client is initialized
    return this.sesClient !== null;
  }
}

// Export singleton instance
export default new EmailService();
